core/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -792,6 +793,17 @@ private void parseSource(DefaultSearchContext context, SearchSourceBuilder sourc
             }
             context.storedFieldsContext(source.storedFields());
         }
+
+        if (source.collapse() != null) {
```

Comment:
I have no strong feelings too about it. It was just a thought stuck in my mind. I'm ok keeping it this way.